PrestaShop/PrestaShop
# Load modules built in translation catalogues

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yes
| Fixed ticket?     | Partially Fixes #19242
| How to test?      | See below
| Possible impacts? | -

## Description

Search and load XLF files in the module translations directory.

The files loaded are in directories `modules/MODULENAME/translations/LOCALENAME`.

The steps during development and the pains encountered :
1. The translation system must load XLF module translations and *when they are not found*, load the core module's translations (= XLF files from the Core that can be used in the module). Indeed it's possible that the Core provides translations for wordings in the module that the modules do not has!
2. Make the translator service load module translations. We have to do it in *TranslatorLanguageLoader*
The problem was that we have multiple ModuleRepository services to retrieve activated modules : The kernel one, the Addon one and one under PrestashopBundle/Entity. There was duplication of logic.
_We cannot modify the last 2 ones as they are used for a very specific usage._
So we needed to use **Kernel/ModuleRepository**.
But this one had too much dependencies to use it everywhere (installer, kernel, ...) so we created a new ModuleRepository in Adapter and use `\Db` to make the query. This new ModuleRepository replaced the Kernel/ModuleRepository.
3. Load module's translations when it is *enabled* (2 cases : module that has been installed just now and module that was previously disabled).
4. The last issue encountered is during install.
The translator is needed in the installer, and that requires the TranslatorLanguageLoader who use ModuleRepository. In the installer, at some steps; the _DB_PREFIX_ constant can be not yet defined and SQL tables can be missing. We dealt with that cases by making the code resilient, able to run even when these elements are not available.

## How to test

Install a module that uses the new translation system (XLF files) (example [xlftranslatedmodule.zip](https://github.com/PrestaShop/PrestaShop/files/6335926/xlftranslatedmodule.zip)). The XLF files must be in the directory `modules/MODULE_NAME/translations/LOCALE/`. These translation files should be acknowledged by the BO Translation interface.

If the module contains XLF files (= new system files) and 'old system translation files' in directory `app/Resources/translations/LOCALE`, the old system translation files are ignored. The new system is used and the old system is ignored.

Also if a module is disabled, you can manually add/modify XLF translation files, and then re-enable it : the translations must be updated.

We can still override these translations by updating them in *Modify translations* page. These overrides are stored in DB

Example of module with translations in it :

[xlftranslatedmodule.zip](https://github.com/PrestaShop/PrestaShop/files/6335926/xlftranslatedmodule.zip)

[translationtest.zip](https://github.com/PrestaShop/PrestaShop/files/6349248/translationtest.zip)


## BC Breaks 

**TranslatorLanguageLoader** definition changes. It now has dependency to **ModuleRepository**

## Deprecations

- PrestaShopBundle\Kernel\ModuleRepositoryFactory
- PrestaShopBundle\Kernel\ModuleRepository


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/24085)
